src/main/java/pl/futurecollars/invoicing/db/file/FileBasedDatabase.java

```diff
@@ -84,14 +87,20 @@ private void updateInvoiceData(Invoice invoice, Invoice data) {
   }
 
   @Override
-  public void delete(int id) {
+  public Optional<Invoice> delete(int id) {
     try {
-      var updatedList = filesService.readAllLines(databasePath)
+      var allInvoices = filesService.readAllLines(databasePath);
+
+      var invoicesExceptDeleted = allInvoices
           .stream()
           .filter(line -> !containsId(line, id))
           .collect(Collectors.toList());
 
-      filesService.writeLinesToFile(databasePath, updatedList);
+      filesService.writeLinesToFile(databasePath, invoicesExceptDeleted);
+
+      allInvoices.removeAll(invoicesExceptDeleted);
+
+      return allInvoices.isEmpty() ? Optional.empty() : Optional.of(jsonService.toObject(allInvoices.get(0), Invoice.class));
```

Comment:
Jak wyżej

Reply:
W dodatku może być tak, że nie będzie istniał obiekt z id podanym w parametrze.
Przeanalizuj jak wtedy zachowa się ta metoda oraz to jak powinna się zachować.

Reply:
To mam ten obiekt zwrócić po id? Nie wiem jak to zrobić. Proszę o jakąś podpowiedź

Reply:
W ogóle nie wiem czy zostawiać tego optionala, czy nie zostać przy void

Reply:
W tym wypadku na spokojnie może być void, ale wolałbym, abyś użył tego voida z wyboru, a nie dlatego, że nie wiesz jak zrobić to inaczej.
W przypadku update'u jak i delete'u jeśli chcemy coś zwracać do mamy kilka sensownych opcji: id, true/false lub cały obiekt (ewentualnie opakowany w optionala). Jeśli zdecydujemy się na id lub cały obiekt, to jedynym sensownym rozwiązaniem jest zwracanie id obiektu, który usunelismy/zaktualizowaliśmy (lub całego obiektu), a nie jakiegoś losowego. W czym mogło by nam to pomóc? Co czego miało by to służyć?

Jeśli chodzi o to skąd wziąć ten obiekt, to już go masz w tej metodzie. Wystarczy go zwrócić :)

Tak jak pisałem na początku: tutaj może być void, ALE mimo to, napisz mi tutaj w komentarzu jak widział byś rozwiązanie ze zwracaniem całego obiektu. W metodzie masz wszystko czego potrzebujesz. Przeanalizuj ją proszę, bo wychodzi na to, że nie do końca wiesz co ona dokładnie robi.

Reply:
public Optional<Invoice> update(int id, Invoice data) {
    try {
      List<String> allInvoices = filesService.readAllLines(databasePath);
      String invoiceAsJson = allInvoices
          .stream()
          .filter(line -> containsId(line, id))
          .findFirst()
          .orElseThrow(() -> new RuntimeException("Invoice with id=" + id + " couldn't be found."));

      allInvoices.remove(invoiceAsJson);
      Invoice invoice = jsonService.toObject(invoiceAsJson, Invoice.class);
      updateInvoiceData(invoice, data);
      allInvoices.add(jsonService.toJson(invoice));
      filesService.writeLinesToFile(databasePath, allInvoices);
      return allInvoices.isEmpty() ? Optional.empty() : Optional.of(data);

    } catch (IOException ex) {
      throw new RuntimeException("Failed to update invoice with id: " + id, ex);
    }

Reply:
Dla update'u zrobiłem coś takiego

Reply:
Nie bardzo ma sens zwracać argumenty metody. My chcemy zwrócić wynik naszych obliczeń, a nie parametr tych obliczeń.
return Optional.of(invoice); 
powinno wystarczyć, ponieważ zmienna invoice zawiera obiekt, który został zaktualizowany